split up ecsql reference into indiviual files, remove iframed iModel … (#6377)
iTwin/itwinjs-core: docs/learning/ECSqlReference/Views.md

# ECViews

Allows defining an abstract entity class which is not backed by actual data but rather by an ECSql query which is executed to obtain instances of the class.
The class is flagged with a custom attribute that is declared in the `ECDbMap` schema and holds the view query ECSql.

The following rules apply to views:

- The custom attribute `ECDbMap.View` must be applied to a standalone abstract entity class.
- View cannot be derived from another class
- View class definition cannot have derived classes.
- All columns selected in the ECSQL query except for ECInstanceId and ECClassId must have corresponding ECProperties defined on the class.
- Query column type and property type have to match (int for integers, string for strings etc.).
- The query has to return an ECInstanceId and an ECClassId column
- All types of properties and computed expressions can be returned by view query if the class definition defines those properties and their types correctly.
- Views can be applied only an ECEntityClass.

> - For relationship classes, a different custom attribute `ECDbMap.ForeignKeyBasedView` can be used instead, which will make the runtime go and check both sides for a navigation property which matches the relationship. If one is found, a view for the relationship is automatically generated.

Reviewer: ```suggestion
- For relationship classes, a different custom attribute `ECDbMap.ForeignKeyBasedView` can be used instead, which will make the runtime check both sides for a navigation property matching the relationship. If one is found, a view for the relationship is automatically generated.
```

> - Any metadata like property category or kind of quantity is taken from the property definitions. Metadata that may come from the view query is overridden by these.

Reviewer: ```suggestion
- Any metadata-like property category or kind of quantity is taken from the property definitions. Metadata that may come from the view query is overridden by these.
```

Author: I refuse to use '-' like this.  It's either one word or two

Reviewer: Alright, well at least here's the justification: 

With `like`, usually, common phrases are one word: _lifelike_, for instance. I'd argue that metadatalike is not a common word. It's also never two words. Otherwise the sentence reads as if you're making a comparison.

                    WHERE json_extract(jo.json, '$.type') = 'pipe'
                </Query>
            </View>
        </ECCustomAttributes>
        <ECProperty propertyName="Diameter" typeName="int" />
        <ECProperty propertyName="Length"  typeName="int"/>
        <ECProperty propertyName="Material" typeName="string" />
    </ECEntityClass>
</ECSchema>
```

`JsonObject` could be filled with this data:

```
{"type": "pipe", "diameter": 10, "length": 100, "material": "steel"}
{"type": "pipe", "diameter": 15, "length": 200, "material": "copper"}
{"type": "pipe", "diameter": 20, "length": 150, "material": "plastic"}
{"type": "cable", "diameter": 5, "length": 500, "material": "copper","type": "coaxial"}
{"type": "cable", "diameter": 2, "length": 1000, "material": "fiber optic","type": "single-mode"}
{"type": "cable", "diameter": 3, "length": 750, "material": "aluminum","type": "twisted pair"}
```

Running this query:

```sql
SELECT Length, Diameter, Material FROM ts.Pipe
```

Will return this result:

```
Length   |Diameter |Material
----------------------------------------
100      |10       |steel
200      |15       |copper
150      |20       |plastic
```

> A view may contain navigation properties, for these, the query can use the `navigation_value`function provided by ECSQL [ECSqlFunctions](./ECSqlFunctions.md)

Reviewer: ```suggestion
A view may contain navigation properties. For these, the query can use the `navigation_value`function provided by ECSQL [ECSqlFunctions](./ECSqlFunctions.md)
```
